components/script/dom/offscreencanvasrenderingcontext2d.rs

```diff
@@ -0,0 +1,951 @@
+use net_traits::image_cache::ImageOrMetadataAvailable;
+use dom::bindings::str::DOMString;
+use net_traits::image_cache::UsePlaceholder;
+use dom::bindings::codegen::UnionTypes::StringOrCanvasGradientOrCanvasPattern;
+use std::{cmp, fmt, mem};
+use dom::imagedata::ImageData;
+use dom::bindings::num::Finite;
+use dom::bindings::codegen::Bindings::CanvasRenderingContext2DBinding::CanvasImageSource;
+use dom::bindings::error::{Error, ErrorResult, Fallible};
+use dom::bindings::codegen::Bindings::CanvasRenderingContext2DBinding::CanvasFillRule;
+use dom::canvasgradient::{CanvasGradient, CanvasGradientStyle};
+use dom::canvaspattern::CanvasPattern;
+use net_traits::image_cache::ImageResponse;
+use net_traits::image::base::PixelFormat;
+use canvas_traits::canvas::{CompositionOrBlending, FillOrStrokeStyle, FillRule};
+use canvas_traits::canvas::{LineCapStyle, LineJoinStyle, LinearGradientStyle};
+use canvas_traits::canvas::{RadialGradientStyle, RepetitionStyle, byte_swap_and_premultiply};
+use dom::bindings::codegen::Bindings::OffscreenCanvasRenderingContext2DBinding;
+use dom::bindings::codegen::Bindings::OffscreenCanvasRenderingContext2DBinding::OffscreenCanvasRenderingContext2DMethods;
+use dom_struct::dom_struct;
+use cssparser::{Parser, ParserInput, RGBA};
+use dom::bindings::reflector::{DomObject, Reflector, reflect_dom_object};
+use ipc_channel::ipc::{self, IpcSender};
+use servo_url::ServoUrl;
+use net_traits::image_cache::ImageCache;
+use euclid::{Transform2D, Point2D, Rect, Size2D};
+use script_traits::ScriptMsg;
+use dom::bindings::root::{Dom, DomRoot, LayoutDom};
+use dom::bindings::codegen::Bindings::ImageDataBinding::ImageDataMethods;
+use std::sync::Arc;
+use std::str::FromStr;
+use dom::globalscope::GlobalScope;
+use dom::node::{document_from_node, window_from_node};
+
+const DEFAULT_WIDTH: u32 = 300;
+const DEFAULT_HEIGHT: u32 = 150;
+
+
+
+
+#[must_root]
+#[derive(Clone, JSTraceable, MallocSizeOf)]
+#[allow(dead_code)]
+enum CanvasFillOrStrokeStyle {
+    Color(RGBA),
+    Gradient(Dom<CanvasGradient>),
+    Pattern(Dom<CanvasPattern>),
+}
+
+#[dom_struct]
+pub struct OffscreenCanvasRenderingContext2D {
+    reflector_: Reflector,
+    #[ignore_malloc_size_of = "Defined in ipc-channel"]
+    ipc_renderer: IpcSender<CanvasMsg>,
+    /// For rendering contexts created by an HTML canvas element, this is Some,
+    /// for ones created by a paint worklet, this is None.
+    canvas: Option<Dom<OffscreenCanvas>>,
+    #[ignore_malloc_size_of = "Arc"]
+    image_cache: Arc<ImageCache>,
+    /// Any missing image URLs.
+    missing_image_urls: DomRefCell<Vec<ServoUrl>>,
+    /// The base URL for resolving CSS image URL values.
+    /// Needed because of https://github.com/servo/servo/issues/17625
+    base_url: ServoUrl,
+    state: DomRefCell<CanvasContextState>,
+    saved_states: DomRefCell<Vec<CanvasContextState>>,
+    origin_clean: Cell<bool>,
+}
+
+#[must_root]
+#[derive(Clone, JSTraceable, MallocSizeOf)]
+struct CanvasContextState {
```

Comment:
We should definitely be able to reuse the CanvasContextState that already exists, rather than duplicating it.

Reply:
Thanks.